sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Rules/PropertiesAccessCorrectField.cs

```diff
@@ -0,0 +1,228 @@
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.CSharp
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    [Rule(DiagnosticId)]
+    public sealed class PropertiesAccessCorrectField : SonarDiagnosticAnalyzer
+    {
+        internal const string DiagnosticId = "S4275";
+        private const string MessageFormat = "Refactor this {0} so that it actually refers to the field '{1}'";
+
+        private static readonly DiagnosticDescriptor rule =
+            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(rule);
+
+        protected override void Initialize(SonarAnalysisContext context)
+        {
+            // We want to check the fields read and assigned in all properties in this class
+            // so this is a symbol-level rule (also means the callback is called only once
+            // for partial classes)
+            context.RegisterSymbolAction(c => CheckType(c),
+                SymbolKind.NamedType);
+        }
+
+        private void CheckType(SymbolAnalysisContext context)
+        {
+            var symbol = (INamedTypeSymbol)context.Symbol;
+            if (symbol.TypeKind != TypeKind.Class &&
+                symbol.TypeKind != TypeKind.Structure)
+            {
+                return;
+            }
+
+            var fields = symbol.GetMembers().Where(m => m.Kind == SymbolKind.Field).OfType<IFieldSymbol>();
+
+            if (!fields.Any())
+            {
+                return;
+            }
+
+            var properties = symbol.GetMembers()
+                .Where(m => m.Kind == SymbolKind.Property)
+                .OfType<IPropertySymbol>()
+                .Where(p => !p.IsImplicitlyDeclared);
+            if (!properties.Any())
+            {
+                return;
+            }
+
+            var fieldToStandardNameMap = new Dictionary<IFieldSymbol, string>();
+            foreach (var field in fields)
+            {
+                fieldToStandardNameMap[field] = GetCanonicalFieldName(field.Name);
+            }
+
+            var allPropertyData = CollectPropertyData(properties, context.Compilation);
+
+            // Check that if there is a single matching field name it is used by the property
+            foreach (var data in allPropertyData)
+            {
+                var matchingFields = fieldToStandardNameMap.Keys.Where(k => AreCanonicalNamesEqual(fieldToStandardNameMap[k], data.CanonicalName));
+                if (matchingFields.Count() == 1)
```

Comment:
If we have multiple candidate field names we give up e.g. 
Fields = a_b and ab,
Property = AB.
We could check that the property at least uses one of the expected fields, but I don't think it's worth it (and it would make the issue message more complicated).